h2/src/main/org/h2/mvstore/tx/TxDecisionMaker.java

```diff
@@ -270,31 +270,25 @@ public final String toString() {
                 long id = existingValue.getOperationId();
                 if (id == 0 // entry is a committed one
                             // or it came from the same transaction
-                        || isThisTransaction(blockingId = TransactionStore.getTransactionId(id))) {
-                    if(existingValue.getCurrentValue() != null) {
-                        return decideToAbort(existingValue.getCurrentValue());
+                        || isThisTransaction(blockingId = TransactionStore.getTransactionId(id))
+                            // or entry belongs to a committing transaction
+                            // and therefore will be committed soon
+                        || isCommitted(blockingId)) {
+                    // check what transaction have been doing and fail it wasn't an entry removal
+                    V currentValue = existingValue.getCurrentValue();
+                    if(currentValue != null) {
+                        return decideToAbort(currentValue);
                     }
-                    if (id == 0) {
+                    if (id == 0 || !isThisTransaction(TransactionStore.getTransactionId(id))) {
```

Comment:
`blockingId` can be reused here.

Reply:
Yep, that's what I think too, but IDEA (and javac) thinks otherwise. It complains that variable may have been undefined. I failed to see how.

Reply:
Yes, that's true, it cannot be uninitialized de-facto, but not in terms of Java language specification. I missed that too.